lib/ProductOpener/Images.pm

```diff
@@ -1349,4 +1349,132 @@ HTML
+
+	if ($ocr_engine eq 'tesseract') {
+
+		my $lan;
+
+		if (defined $ProductOpener::Config::tesseract_ocr_available_languages{$lc}) {
+			$lan = $ProductOpener::Config::tesseract_ocr_available_languages{$lc};
+		}
+		elsif (defined $ProductOpener::Config::tesseract_ocr_available_languages{$product_ref->{lc}}) {
+			$lan = $ProductOpener::Config::tesseract_ocr_available_languages{$product_ref->{lc}};
+		}
+		elsif (defined $ProductOpener::Config::tesseract_ocr_available_languages{en}) {
+			$lan = $ProductOpener::Config::tesseract_ocr_available_languages{en};
+		}
+
+		$log->debug("extracting text with tesseract", { lc => $lc, lan => $lan, id => $id, image => $image }) if $log->is_debug();
+
+		if (defined $lan) {
+			$text =  decode utf8=>get_ocr($image,undef,$lan);
+
+			if ((defined $text) and ($text ne '')) {
+				$product_ref->{$field} = $text;
+				$status = 0;
+			}
+		}
+		else {
+			$log->warn("no available tesseract dictionary", { lc => $lc, lan => $lan, id => $id }) if $log->is_warn();
+		}
+
+	}
+	elsif ($ocr_engine eq 'google_cloud_vision') {
+
+		my $url = "https://alpha-vision.googleapis.com/v1/images:annotate?key=" . $ProductOpener::Config::google_cloud_vision_api_key;
+		# alpha-vision.googleapis.com/
+
+		my $ua = LWP::UserAgent->new();
+
+		my $api_request_ref =
+			{
+				requests =>
+					[
+						{
+							features => [{ type => 'TEXT_DETECTION'}], image => { source => { imageUri => $image_url}}
+						}
+					]
+			}
+		;
+		my $json = encode_json($api_request_ref);
+
+		my $request = HTTP::Request->new(POST => $url);
+		$request->header( 'Content-Type' => 'application/json' );
+		$request->content( $json );
+
+		my $res = $ua->request($request);
+
+		if ($res->is_success) {
+
+			$log->info("request to google cloud vision was successful") if $log->is_info();
+
+			my $json_response = $res->decoded_content;
+
+			my $cloudvision_ref = decode_json($json_response);
+
+			my $json_file = "$www_root/images/products/$path/$filename.full.jpg" . ".google_cloud_vision.json";
+
+			$log->info("saving google cloud vision json response to file", { path => $json_file }) if $log->is_info();
+
+			# UTF-8 issue , see https://stackoverflow.com/questions/4572007/perl-lwpuseragent-mishandling-utf-8-response
+			$json_response = decode("utf8", $json_response);
+
+			open (my $OUT, ">:encoding(UTF-8)", $json_file);
+			print $OUT $json_response;
+			close $OUT;
+
+			if ((defined $cloudvision_ref->{responses}) and (defined $cloudvision_ref->{responses}[0])
+				and (defined $cloudvision_ref->{responses}[0]{fullTextAnnotation})
+				and (defined $cloudvision_ref->{responses}[0]{fullTextAnnotation}{text})) {
+
+				$log->debug("text found in google cloud vision response") if $log->is_debug();
+
+
+				$product_ref->{$field} = $cloudvision_ref->{responses}[0]{fullTextAnnotation}{text};
+				$product_ref->{$field . "_annotations"} = $cloudvision_ref;
```

Comment:
This saves the complete cloudvision response in the `product.sto`, right?

1. This could be a lot of data to be stored in the database and to be returned in our API (large API responses)?
2. By saving it in our database, we would basically publish the cloudvision data under the ODbL, and mixing ODbL with other sources is weird. Do the terms of service of the Cloud Vision API allow this?

Reply:
Nope, it saves on disk.

Reply:
> Nope, it saves on disk.

Not necessarily. If `extract_text_from_image` was called with a product reference and that reference was then stored on disk as usual, it would (at some point) be stored in MongoDB, too. If that's not intentional, then it would be weird to use the `$product_ref` only as a container for the resulting data, because that's not how the reference is used throughout ProductOpener.

Reply:
It's only called from ingredients.pl, which does not save the product. I did not intend the annotations to be saved in product.sto or mongodb. It won't do it at this point, but we could pass an extra parameter for the resulting data (both the text and the annotations).

Reply:
Now using a separate $results_ref structure for results.